[feat] 상품 상세 조회시 조회수 증가

pie2457 opened the pull request:
## Issues
- #50 

## What is this PR? 👓
상품 상세 조회시 조회수 증가에 대한 PR입니다.

## Key changes 🔑
- 상품 상세조회 요청시 viewCount 증가
- 상품 상세조회 요청시 viewCount 증가 테스트
- 이미 같은 상품을 조회했던 사용자가 또 다시 상품 상세조회 요청을 보냈을 시 뷰카운트 증가는 하지 않습니다.
- 상품의 상세 조회 페이지는 접근이 빈번하게 일어날 것이라고 예상해,
IO 연산을 줄이기 위해 들어온 요청을 캐싱해두고 스케쥴링 시간에 맞춰 RDB에 저장해주는 방식을 구현했습니다.

pie2457 committed:
docs #50 : redis 사용을 위한 dependency 추가

pie2457 committed:
feat #50 : 상품 상세 조회 시 viewCount 증가

이미 같은 상품을 조회했던 사용자가 또 다시 상품 상세조회 요청을 보냈을 시 뷰카운트 증가는 하지 않는다.

pie2457 committed:
test #50 : 상품 상세 조회 시 viewCount 증가 테스트

- 사용되지 않는 변수명 제거

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/config/RedisConfig.java:
```diff
@@ -0,0 +1,34 @@
+
+@Configuration
+@RequiredArgsConstructor
+@EnableRedisRepositories
```
`RedisRepository`를 사용하지 않아 해당 애노테이션은 필요 없을 것 같네요~

pie2457 replied:
헉 그렇네요 감사합니다 !!

23Yong commented on src/main/java/pie/tomato/tomatomarket/infrastructure/config/RedisConfig.java:
```diff
@@ -0,0 +1,34 @@
+
+	@Bean
+	public RedisConnectionFactory redisConnectionFactory() {
+		return new LettuceConnectionFactory(redisProperties.getHost(), redisProperties.getPort());
```
오 Lettuce 를 사용하셨군용
Lettuce 말고 Jedis, Redisson 이라는 것도 있습니다. 궁금하시면 찾아봐도 좋을 것 같아요

pie2457 replied:
jedis는 deprecated 됐다는걸 봤는데 redission은 처음보네요 감사합니다 찾아볼게요 ㅎㅎ

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/item/ItemService.java:
```diff
@@ -168,7 +170,8 @@ private void verifyExistsItem(Long itemId) {
 
-	public ItemDetailResponse itemDetails(Long itemId) {
+	public ItemDetailResponse itemDetails(Principal principal, Long itemId) {
+		itemViewCountRedisService.addViewCount(principal.getNickname(), itemId);
 		Item findItem = itemRepository.findById(itemId)
 			.orElseThrow(() -> new NotFoundException(ErrorCode.NOT_FOUND_ITEM));
```
두 로직의 순서가 바뀌어야 할 것 같아요
찾고자하는 아이템이 없을 때 예외를 던지고, 있다면 조회수를 증가시키는게 좋겠죠?

pie2457 replied:
네 그게 좋겠네요 !! 감사합니당

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+@RequiredArgsConstructor
+public class ItemViewCountRedisService {
+
+	public static final String ITEM_ID_PREFIX = "itemId: ";
```
여기는 접근제어자가 `public` 이네요

23Yong replied:
그리구 `itemId: ` 도 좋지만 `itemViewCount: `는 어떠세요?
prefix가 어떤 것에 대한 키인지 알 수 있으면 좋을 것 같아서요!

pie2457 replied:
접근제어자는 확인을 못했네요ㅠㅠ 감사합니다 수정하겠습니다 !
제가 생각한 `Key : value` 방식은 key에 `"itemId : " + itemId`를 담고 value값에 `viewCount`에 대한 값을 넣으려 생각한것인데 prefix가 어떤 것에 대한 키인지 명확히 하려면 생각을 좀 바꿔봐야겠네요 좋은 의견 감사합니다 :) !

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+
+	public void addViewCount(String nickname, Long itemId) {
+		String key = ITEM_ID_PREFIX + itemId;
+		String duplicateKey = nickname;
```
key는 조회수와 관련된 상품의 pk를, duplicateKey가 사용자의 nickname을 의미하는군요

그렇다면 itemViewCountKey, nicknameKey 와 같은 네임은 어떠세요?

pie2457 replied:
네 그게 좋겠네요 감사합니다 ㅎ.ㅎ

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+	public void addViewCount(String nickname, Long itemId) {
+		String key = ITEM_ID_PREFIX + itemId;
+		String duplicateKey = nickname;
+		ValueOperations<String, String> value = redisTemplate.opsForValue();
```
해당 변수 선언이 36라인으로 가면 좋을 것 같아요~

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	private boolean isNotFirstView(String duplicateKey) {
+		return Boolean.TRUE.equals(redisTemplate.hasKey(duplicateKey));
```
```java
return redisTemplate.hasKey(duplicateKey);
```

이렇게 하지 않은 이유가 있나요?

pie2457 replied:
아뇨 생각이 모자랐나봅니다.. ㅎㅎㅎ ㅎㅎ 제가 부족한 점이 많네요 ,, ㅠㅠ 감사합니다 !!

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+			return;
+		}
+
+		value.set(duplicateKey, String.valueOf(itemId));
```
이렇게 하면 동일한 회원이 다른 상품에 접근할 때 문제가 되지 않나요?
`isNotFirstView` 로직에 걸릴 것 같아서요!

pie2457 replied:
key값으로만 검사하니까 그렇겠네요 value에 itemId가 있어서 거기까진 생각못했네요ㅠㅠ 감사합니다 ! 방법을 바꿔야겠네요 !!

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+			Long itemId = Long.parseLong(key.split(" ")[1]);
+			Long viewCount = Long.valueOf(Optional.ofNullable(redisTemplate.opsForValue().get(key)).orElse("0"));
+			Long originViewCount = itemRepository.findViewCountById(itemId);
+			itemRepository.addViewCountFromRedis(itemId, originViewCount + viewCount);
```
item을 조회해와서 JPA의 더티체킹을 이용해도 되고~ 
지금처럼 해도 될 것 같네요~

pie2457 replied:
헉 그렇겠네요  감사합니다 ㅎㅎ

23Yong commented on src/main/java/pie/tomato/tomatomarket/application/redis/ItemViewCountRedisService.java:
```diff
@@ -0,0 +1,65 @@
+			Long originViewCount = itemRepository.findViewCountById(itemId);
+			itemRepository.addViewCountFromRedis(itemId, originViewCount + viewCount);
+			redisTemplate.delete(key);
+			redisTemplate.delete("itemId: " + itemId);
```
여기는 `ITEM_ID_PREFIX`를 써도 될 것 같네요~

pie2457 replied:
앗 그러네요 감사합니다 ㅎㅎㅎ

23Yong reviewed (approved):
레디스 어렵군요.. 수고하셨습니다~!

pie2457 committed:
refactor #50 : 리뷰사항 반영

- RedisConfig : 사용하지 않는 애노테이션 제거
- ItemService : itemId가 정상적으로 들어왔는지 확인 후 뷰카운트 증가, 가독성을 위한 공백 추가
- Item : 더티체킹을 위한 메서드 추가
- ItemRepository : 코드 수정으로 사용하지 않는 메서드 제거
- ItemViewCountRedisService :
  - ITEM_ID_PREFIX - > ITEM_VIEW_COUNT_ID_PREFIX : 명확하지 않은 상수 이름 변경
  - key, duplicateKey -> itemViewCountKey, nicknameKey : key의 사용성에 따라 변수 이름 변경
  - nicknameKey : 중복 검증 방식이 nickname으로 되어있어서 다른 item을 조회하더라도 viewCount가 올라가지 않음 nickname + itemId로 변경
  - isNotFirstView : 굳이 필요없는 Boolean.TRUE.equals() 제거
- deleteViewCountCache() :
  - 더티체킹으로 viewCount를 업데이트하도록 변경
